src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -106,5 +116,95 @@ private IArgumentConstraint CreateParamArrayConstraint(NewArrayExpression expres
 
             return new AggregateArgumentConstraint(result);
         }
+
+        private class ArgumentConstraintExpressionVisitor : ExpressionVisitor
+        {
+            private readonly List<Expression> constraintExpressions;
+            private readonly HashSet<Expression> visitedExpressions;

```

Comment:
I did, but I'm not so sure now. It may be someotherorder traversal, but you double visit because of the shenanigans that I didn't entirely understand - you're forcing a descent down the tree in some cases.